### PR TITLE
XML Source Endpoint

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Make it possible to represent page source as XML document. This should simplify XPaths queries debugging and will make the life much easier for Appium devs. 

@imurchie FYI

### CODE REVIEW

File: WebDriverAgentLib/Commands/FBDebugCommands.m [hunk 1]
```diff
@@ -43,4 +46,25 @@ + (NSArray *)routes
   return FBResponseWithStatus(FBCommandStatusNoError, @{ @"tree": (accessibleTreeType ? application.fb_accessibilityTree : application.fb_tree) ?: @{} } );
 }
 
++ (id<FBResponsePayload>)handleGetSourceCommand:(FBRouteRequest *)request
+{
+  FBApplication *application = request.session.application ?: [FBApplication fb_activeApplication];
+  if (!application) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: This will never happen anymore.

File: WebDriverAgentLib/Commands/FBDebugCommands.m [hunk 2]
```diff
@@ -43,4 +46,25 @@ + (NSArray *)routes
...
+    return FBResponseWithErrorFormat(@"There is no active application");
+  }
+  NSString *sourceType = request.parameters[@"type"];
+  id result;
+  if ([sourceType caseInsensitiveCompare:@"json"] == NSOrderedSame) {
+    result = application.fb_tree;
+  } else if ([sourceType caseInsensitiveCompare:@"xml"] == NSOrderedSame) {
+    result = [FBXPath getSnapshotAsXMLString:application.lastSnapshot];
+    if (nil == result) {
```
Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: You could share result check and move it to line 67

File: WebDriverAgentLib/Commands/FBDebugCommands.m (hunk 1 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: removed verification. Shall I add NSAssert instead?

File: WebDriverAgentLib/Commands/FBDebugCommands.m (hunk 2 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: done

File: WebDriverAgentLib/Commands/FBDebugCommands.m (hunk 1 above)
Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: Good idea. I just realised that there is one case it can happen. If you start query app that hosts WDA.

Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: Or instead of adding asserts for that everywhere we could handle that in FBapplication

File: WebDriverAgentLib/Commands/FBDebugCommands.m [hunk 3]
```diff
@@ -22,6 +23,9 @@ + (NSArray *)routes
 {
   return
   @[
+    [[FBRoute GET:@"/source/:type"] respondWithTarget:self action:@selector(handleGetSourceCommand:)],
+    [[FBRoute GET:@"/source/:type"].withoutSession respondWithTarget:self action:@selector(handleGetSourceCommand:)],
+    // TODO: Kill source methods, that support json representation only
```
Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: If you want to kill other endpoints you should patch Inspector as well.

File: WebDriverAgentLib/Commands/FBDebugCommands.m [hunk 4]
```diff
@@ -43,4 +47,25 @@ + (NSArray *)routes
...
+  if ([sourceType caseInsensitiveCompare:@"json"] == NSOrderedSame) {
+    result = application.fb_tree;
+  } else if ([sourceType caseInsensitiveCompare:@"xml"] == NSOrderedSame) {
+    result = [FBXPath getSnapshotAsXMLString:application.lastSnapshot];
+  } else {
+    status = FBCommandStatusUnsupported;
+    result = [NSString stringWithFormat:@"Unknown source type '%@'. Only 'xml' and 'json' source types are supported.", sourceType];
+  }
+  if (!result) {
+    status = FBCommandStatusUnhandled;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: I would keep `FBCommandStatusUnsupported` & `FBCommandStatusUnhandled ` as fast returns otherwise it is hard to read.

File: WebDriverAgentLib/Utilities/FBXPath.h [hunk 5]
```diff
@@ -50,6 +50,15 @@ extern NSString *const XCElementSnapshotXPathQueryEvaluationException;
...
 + (NSArray<XCElementSnapshot *> *)findMatchesIn:(XCElementSnapshot *)root xpathQuery:(NSString *)xpathQuery;
 
+/**
+ Gets XML representation of XCElementSnapshot with all its descendants. This method generates the same
+ representation, which is used for XPath search
+ 
+ @param root the root element
+ @return valid XML document as string or nil in case of failure
+ */
++ (nullable NSString *)getSnapshotAsXMLString:(XCElementSnapshot *)root;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBXPath.h]: Drop `get` it is not objc like. Maybe `XMLStringWithSnapshot:`?

File: WebDriverAgentLib/Commands/FBDebugCommands.m (hunk 3 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: this should be another PR. Also, we need to sync with Appium, since they use this endpoint as well. Whet we can do for now - use the single _ handleGetSourceCommand_ method for all endpoints to avoid code duplication

File: WebDriverAgentLib/Commands/FBDebugCommands.m (hunk 4 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBDebugCommands.m]: OK

File: WebDriverAgentLib/Utilities/FBXPath.h (hunk 5 above)
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBXPath.h]: np

File: WebDriverAgentLib/Utilities/FBXPath.h [hunk 6]
```diff
@@ -50,6 +50,15 @@ extern NSString *const XCElementSnapshotXPathQueryEvaluationException;
...
 + (NSArray<XCElementSnapshot *> *)findMatchesIn:(XCElementSnapshot *)root xpathQuery:(NSString *)xpathQuery;
 
+/**
+ Gets XML representation of XCElementSnapshot with all its descendants. This method generates the same
+ representation, which is used for XPath search
+ 
+ @param root the root element
+ @return valid XML document as string or nil in case of failure
+ */
++ (nullable NSString *)xmlStringWithSnapshot:(XCElementSnapshot *)root;
```
Comment (marekcirkos) [inline WebDriverAgentLib/Utilities/FBXPath.h]: Add tests for that method

Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBXPath.h]: There is a test in FBXPathIntegrationTests->testSingleDescendantXMLRepresentation

**Comment by marekcirkos** [comment]
Thanks @mykola-mokhnach for your help with WDA!

**Comment by marekcirkos** [comment]
@mykola-mokhnach You are good to rebase. I noticed that testInvisibleDescendantWithXPathQuery is consistently failing. Can you fix that preferably on different diff?

**Comment by mykola-mokhnach** [comment]
@marekcirkos It fails because of the good old https://github.com/facebook/WebDriverAgent/issues/372

```
Enqueue Failure: UI Testing Failure - Failure fetching attributes for element <XCAccessib
ilityElement: 0x61000024c510> Device element: Error Domain=XCTestManagerErrorDomain Code=13 "Error copying attributes -25202" UserInfo={NSLocalizedDescription=Error copying attributes -25202} <unknown> 0 1
```

issue, which is also visible in Travis logs. It happens, because running it on Travis is slower, than on local machine, so there is some transition animation in progress while WDA is trying to read _wdVisible_ property. This [PR](https://github.com/facebook/WebDriverAgent/pull/369), that you closed previously, was supposed to fix it by explicitly waiting until transition animation is finished. So now I can only add a hardcoded sleep and hope, that Travis is fast enough to finish the animation before the sleep expires.

**Comment by mykola-mokhnach** [comment]
@marekcirkos I'd rather create a new PR with all these changes. I hate rebase.

**Comment by mykola-mokhnach** [comment]
The thing is I have already merged it with master. So I don't know a way to rebase the current branch now without creating a new branch containing all the changes (I can simply sync it with the current branch by using **git read-tree** utility).
